core/store/src/lib.rs

```diff
@@ -44,11 +44,11 @@ mod trie;
 
 #[derive(Clone)]
```

Comment:
@mm-near this feels like an important type to document. I am especially wondering about our threading strategy here -- this type is `Clone`, so many threads may access the database at the same time, and they can also modify it concurrently via `Store::store_update`. And then there are things like iterators which, at least judging from the types, can happen concurrently with modification.

Reply:
I can comment on how store is used in the chain and runtime level and @pmnoxx can comment on how it is accessed in `PeerManagerActor` and other network related threads. My guess is that the network threads access completely different database columns than chain and runtime would access, so they don't conflict with each other.

The general story is that although `Store` can be modified concurrently, in practice we don't do that. In chain and runtime, `Store` is used by three structs, `Chain`, `NightshadeRuntime` and `EpochManager`. `Chain` is accessed by one thread (`ClientActor`) while `NightshadeRuntime` and `EpochManager` are accessed by multiple threads (`ViewClientActor`, `ClientActor` and temporary threads when we apply blocks). The trick is, only `Chain` actually modifies store through `store_update`, `EpochManager` and `NightshadeRuntime` only use `Store` to read information, they never write (except for maybe during initialization).